src/github.com/mozilla-services/pushgo/simplepush/handlers_endpoint.go

```diff
@@ -123,21 +128,94 @@ func (h *EndpointHandler) Start(errChan chan<- error) {
 	errChan <- h.server.Serve(h.listener)
 }
 
+func (h *EndpointHandler) decodePK(token string) (key string, err error) {
+	if len(token) == 0 {
+		return "", fmt.Errorf("Missing primary key")
+	}
+	if len(h.tokenKey) == 0 {
+		return token, nil
+	}
+	bpk, err := Decode(h.tokenKey, token)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes.TrimSpace(bpk)), nil
+}
+
+func (h *EndpointHandler) resolvePK(token string) (uaid, chid string, err error) {
+	pk, err := h.decodePK(token)
+	if err != nil {
+		err = fmt.Errorf("Error decoding primary key: %s", err)
+		return "", "", err
+	}
+	if !validPK(pk) {
+		err = fmt.Errorf("Invalid primary key: %q", pk)
+		return "", "", err
+	}
+	uaid, chid, ok := h.store.KeyToIDs(pk)
+	if !ok {
+		err = fmt.Errorf("Could not resolve primary key: %q", pk)
+		return "", "", err
+	}
+	if len(chid) == 0 {
+		err = fmt.Errorf("Primary key missing channel ID: %q", pk)
+		return "", "", err
+	}
+	return uaid, chid, nil
+}
+
+func (h *EndpointHandler) doPropPing(uaid string, version int64, data string) (ok bool, err error) {
+	if h.pinger == nil {
+		return false, nil
+	}
+	if ok, err = h.pinger.Send(uaid, version, data); err != nil {
+		return false, fmt.Errorf("Could not send proprietary ping: %s", err)
+	}
+	if !ok {
+		return false, nil
+	}
+	/* if this is a GCM connected host, boot vers immediately to GCM
+	 */
+	return h.pinger.CanBypassWebsocket(), nil
+}
+
+// getUpdateParams extracts the update version and data from req.
+func (h *EndpointHandler) getUpdateParams(req *http.Request) (version int64, data string, err error) {
```

Comment:
I wonder if we should return a struct containing the Update parameters rather than each discretely. Not a big priority, since we currently only have two, but if we start adding things, this could get a bit long.